### PR TITLE
[SimplifyCFG] Select the first instruction that we can handle in `passingValueIsAlwaysUndefined`

### LINKED ISSUE #98799: [SimplifyCFG] SimplifyCFG does not eliminate the UB branch after inlining
I tried the following IR:

```
define internal ptr @bar(ptr %arg, i1 %arg1) {
bb:
  br i1 %arg1, label %bb4, label %bb2

bb2:
  %i = load ptr, ptr %arg, align 8
  %i3 = getelementptr inbounds i8, ptr %i, i64 1
  store ptr %i3, ptr %arg, align 8
  br label %bb4

bb4:
  %i5 = phi ptr [ %i, %bb2 ], [ null, %bb ]
  ret ptr %i5
}

define i32 @foo(ptr %arg, i1 %arg1) {
bb:
  %i = call ptr @bar(ptr %arg, i1 %arg1)
  %i2 = icmp ne ptr %i, null
  call void @llvm.assume(i1 %i2)
  %i3 = load i32, ptr %i, align 4
  ret i32 %i3
}

declare void @llvm.assume(i1)
```

We can eliminate the branch, but it doesn't. To my surprise, re-executing the output once simplifycfg eliminates this branch: https://llvm.godbolt.org/z/85zKbMdxe.

I've noticed that the user order is different after inlining and reading the text IR.

The `passingValueIsAlwaysUndefined` only considers the first user:
https://github.com/llvm/llvm-project/blob/92f4001906a18fca29929a333e61fdd662a9b0bd/llvm/lib/Transforms/Utils/SimplifyCFG.cpp#L7567-L7577

Of course there are many ways to solve this problem, for example:

1. Handle all users
2. Choose an instruction that `passingValueIsAlwaysUndefined` can handle (I'd go with this one first.)
3. Support `assume(%p == null)`

But my biggest concern here is **whether the user order needs to be consistent**. I don't like it when an optimization happens randomly.

### PR DESCRIPTION
**Pull request by dianqk**
Fixes #98799.

Perf: https://llvm-compile-time-tracker.com/compare.php?from=181e4c6291c94a38c0ee89d2128f8d70b15d2d23&to=b686e35ecd81ae76fc517d14e133d44ed7ff9f4b&stat=instructions%3Au

### CODE REVIEW

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 1]
```diff
@@ -7573,13 +7573,25 @@ static bool passingValueIsAlwaysUndefined(Value *V, Instruction *I, bool PtrValu
     return false;
 
   if (C->isNullValue() || isa<UndefValue>(C)) {
-    // Only look at the first use, avoid hurting compile time with long uselists
-    auto *Use = cast<Instruction>(*I->user_begin());
-    // Bail out if Use is not in the same BB as I or Use == I or Use comes
-    // before I in the block. The latter two can be the case if Use is a PHI
-    // node.
-    if (Use->getParent() != I->getParent() || Use == I || Use->comesBefore(I))
+    // Only look at the first use we can hanle, avoid hurting compile time with
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:7576]:
```suggestion
    // Only look at the first use we can handle, avoid hurting compile time with
```

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 2]
```diff
@@ -7573,13 +7573,25 @@ static bool passingValueIsAlwaysUndefined(Value *V, Instruction *I, bool PtrValu
...
+      // before I in the block. The latter two can be the case if Use is a
+      // PHI node.
+      if (Use->getParent() != I->getParent() || Use == I || Use->comesBefore(I))
+        return false;
+      // Change this list when we want to add new instructions.
+      if (!isa<GetElementPtrInst>(Use) && !isa<ReturnInst>(Use) &&
+          !isa<BitCastInst>(Use) && !isa<LoadInst>(Use) &&
+          !isa<StoreInst>(Use) && !isa<AssumeInst>(Use) && !isa<CallBase>(Use))
+        return false;
+      return true;
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:7590]: Can you turn this into a switch?

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 3]
```diff
@@ -7573,13 +7573,36 @@ static bool passingValueIsAlwaysUndefined(Value *V, Instruction *I, bool PtrValu
...
+      case Instruction::Ret:
+        return true;
+      case Instruction::BitCast:
+        return true;
+      case Instruction::Load:
+        return true;
+      case Instruction::Store:
+        return true;
+      case Instruction::Call:
+        return true;
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:7600]:
```suggestion
      case Instruction::GetElementPtr:
      case Instruction::Ret:
      case Instruction::BitCast:
      case Instruction::Load:
      case Instruction::Store:
      case Instruction::Call:
      case Instruction::Invoke:
      case Instruction::CallBr:
        return true;
```

**Comment by dtcxzyw** [comment]
IR diff LGTM. I am running compile-time tests :)

BTW, I think most of dead blocks are created by JumpThreading. Can we avoid this to further improve compilation time?

**Comment by DianQK** [comment]
> BTW, I think most of dead blocks are created by JumpThreading. Can we avoid this to further improve compilation time?

I haven't read the JumpThreading code yet, if it can't be handled within it, perhaps we could create some sort of simple matching pattern to hand over to SimplifyCFG?

**Comment by DianQK** [comment]
IMO, checking all the uses works too: https://llvm-compile-time-tracker.com/compare.php?from=181e4c6291c94a38c0ee89d2128f8d70b15d2d23&to=1d88f9270188c0ebc8bc89f1cee6a9f5c2045974&stat=instructions:u.

**Comment by dtcxzyw** [comment]
> IR diff LGTM. I am running compile-time tests :)
> 

```
Top 5 improvements:
  yosys/opt_muxtree.ll 6478612386 5814367885 -10.25%
  proxy/proxy_lifetime_tests.cpp.ll 7675508865 6910976274 -9.96%
  luau/InsertionOrderedMap.test.cpp.ll 1564039427 1431728556 -8.46%
  yosys/extract_reduce.ll 6042277718 5548421964 -8.17%
  yosys/attrmvcp.ll 3867732946 3646575352 -5.72%
Top 5 regressions:
  cmake/cmGeneratorExpressionLexer.cxx.ll 280987017 355193810 +26.41%
  hermes/StackPromotion.cpp.ll 5084729293 5932637578 +16.68%
  verilator/V3GraphPathChecker.cpp.ll 384704949 442379033 +14.99%
  yosys/shregmap.ll 6984227787 7798033142 +11.65%
  assimp/Exporter.cpp.ll 2987022380 3312585333 +10.90%
Overall: 0.01310176%
```

**Comment by DianQK** [comment]
I can't reproduce the problem yet, can you provide a more detailed test step? 
My local test commands are:

```bash
# CMake project
clang++     -D_FILE_OFFSET_BITS=64    -DCMAKE_BOOTSTRAP    -DCMake_HAVE_CXX_MAKE_UNIQUE=1 -DCMake_HAVE_CXX_FILESYSTEM=1   -I/home/dianqk/llvm/CMake/build/Bootstrap.cmk   -I/home/dianqk/llvm/CMake/Source   -I/home/dianqk/llvm/CMake/Source/LexerParser   -I/home/dianqk/llvm/CMake/Utilities/std   -I/home/dianqk/llvm/CMake/Utilities  -c /home/dianqk/llvm/CMake/Source/cmGeneratorExpressionLexer.cxx -o cmGeneratorExpressionLexer.o -emit-llvm -O3 -Xclang -disable-llvm-passes

# cmake (stage 1)
cmake -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=OFF -DBUILD_SHARED_LIBS=ON ...
# main branch
perf stat $LLVM_PROJECT/build-base/bin/opt -O2 cmGeneratorExpressionLexer.o --disable-output
339,409,679      instructions:u
# pr
perf stat $LLVM_PROJECT/build/bin/opt -O2 cmGeneratorExpressionLexer.o --disable-output
338,379,357      instructions:u
```

**Comment by DianQK** [comment]
It's surprising that https://github.com/llvm/llvm-project/pull/98802/commits/b686e35ecd81ae76fc517d14e133d44ed7ff9f4b is making things worse: https://llvm-compile-time-tracker.com/compare.php?from=057abc842e581d01ac0bff435d779e3f506440ad&to=b686e35ecd81ae76fc517d14e133d44ed7ff9f4b&stat=instructions:u.

**Comment by dtcxzyw** [comment]
> I can't reproduce the problem yet, can you provide a more detailed test step? My local test commands are:
> 
> ```shell
> # CMake project
> clang++     -D_FILE_OFFSET_BITS=64    -DCMAKE_BOOTSTRAP    -DCMake_HAVE_CXX_MAKE_UNIQUE=1 -DCMake_HAVE_CXX_FILESYSTEM=1   -I/home/dianqk/llvm/CMake/build/Bootstrap.cmk   -I/home/dianqk/llvm/CMake/Source   -I/home/dianqk/llvm/CMake/Source/LexerParser   -I/home/dianqk/llvm/CMake/Utilities/std   -I/home/dianqk/llvm/CMake/Utilities  -c /home/dianqk/llvm/CMake/Source/cmGeneratorExpressionLexer.cxx -o cmGeneratorExpressionLexer.o -emit-llvm -O3 -Xclang -disable-llvm-passes
> 
> # cmake (stage 1)
> cmake -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=OFF -DBUILD_SHARED_LIBS=ON ...
> # main branch
> perf stat $LLVM_PROJECT/build-base/bin/opt -O2 cmGeneratorExpressionLexer.o --disable-output
> 339,409,679      instructions:u
> # pr
> perf stat $LLVM_PROJECT/build/bin/opt -O2 cmGeneratorExpressionLexer.o --disable-output
> 338,379,357      instructions:u
> # b686e35ec
> 339,831,704      instructions:u
> ```

Can you try with `opt -O3`?

**Comment by DianQK** [comment]
Oops, I can reproduce it. b686e35ecd81ae76fc517d14e133d44ed7ff9f4b fixed.

**Comment by dtcxzyw** [comment]
> Oops, I can reproduce it. [b686e35](https://github.com/llvm/llvm-project/commit/b686e35ecd81ae76fc517d14e133d44ed7ff9f4b) fixed.

```
Top 5 improvements:
  faiss/IndexIVFFlat.cpp.ll 1879640556 1806714892 -3.88%
  faiss/utils.cpp.ll 1392144764 1341384014 -3.65%
  faiss/IndexIVFFastScan.cpp.ll 7591452840 7321469207 -3.56%
  gromacs/statepropagatordata.cpp.ll 3289712205 3176646034 -3.44%
  rust-analyzer-rs/5ac99zaxn7b9r9xv.ll 2542407788 2477356558 -2.56%
Top 5 regressions:
  faiss/IVFlib.cpp.ll 1451629900 1522431580 +4.88%
  gromacs/constr.cpp.ll 2086259177 2159075037 +3.49%
  gromacs/muParserBase.cpp.ll 3535512461 3655648190 +3.40%
  faiss/IndexIVFAdditiveQuantizer.cpp.ll 758961920 784502318 +3.37%
  faiss/IndexAdditiveQuantizer.cpp.ll 1739627273 1791928381 +3.01%
Overall: 0.00096778%
```

**Comment by DianQK** [comment]
Hmm, at least locally, I can't reproduce `gromacs/constr.cpp`. The new overall results seem acceptable?

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**